# Practica SQL

In [1]:
# Importamos la libreria para poder trabajar 
import sqlite3

### 1. Conectamos con la librería 

In [2]:
conexion = sqlite3.connect('Chinook_Sqlite.sqlite')

### 2. Creamos un cursor

In [3]:
cursor = conexion.cursor()

### 3. Trabajamos con la tabla Customer

**3.1 Mostramos la columna email**

In [4]:
cursor.execute("SELECT Email FROM Customer")
email = cursor.fetchall()
email

[('luisg@embraer.com.br',),
 ('leonekohler@surfeu.de',),
 ('ftremblay@gmail.com',),
 ('bjorn.hansen@yahoo.no',),
 ('frantisekw@jetbrains.com',),
 ('hholy@gmail.com',),
 ('astrid.gruber@apple.at',),
 ('daan_peeters@apple.be',),
 ('kara.nielsen@jubii.dk',),
 ('eduardo@woodstock.com.br',),
 ('alero@uol.com.br',),
 ('roberto.almeida@riotur.gov.br',),
 ('fernadaramos4@uol.com.br',),
 ('mphilips12@shaw.ca',),
 ('jenniferp@rogers.ca',),
 ('fharris@google.com',),
 ('jacksmith@microsoft.com',),
 ('michelleb@aol.com',),
 ('tgoyer@apple.com',),
 ('dmiller@comcast.com',),
 ('kachase@hotmail.com',),
 ('hleacock@gmail.com',),
 ('johngordon22@yahoo.com',),
 ('fralston@gmail.com',),
 ('vstevens@yahoo.com',),
 ('ricunningham@hotmail.com',),
 ('patrick.gray@aol.com',),
 ('jubarnett@gmail.com',),
 ('robbrown@shaw.ca',),
 ('edfrancis@yachoo.ca',),
 ('marthasilk@gmail.com',),
 ('aaronmitchell@yahoo.ca',),
 ('ellie.sullivan@shaw.ca',),
 ('jfernandes@yahoo.pt',),
 ('masampaio@sapo.pt',),
 ('hannah.schneider@

**3.2 Mostramos la columna FirstName**

In [5]:
cursor.execute("SELECT FirstName FROM Customer")
firstName = cursor.fetchall()
firstName


[('Luís',),
 ('Leonie',),
 ('François',),
 ('Bjørn',),
 ('František',),
 ('Helena',),
 ('Astrid',),
 ('Daan',),
 ('Kara',),
 ('Eduardo',),
 ('Alexandre',),
 ('Roberto',),
 ('Fernanda',),
 ('Mark',),
 ('Jennifer',),
 ('Frank',),
 ('Jack',),
 ('Michelle',),
 ('Tim',),
 ('Dan',),
 ('Kathy',),
 ('Heather',),
 ('John',),
 ('Frank',),
 ('Victor',),
 ('Richard',),
 ('Patrick',),
 ('Julia',),
 ('Robert',),
 ('Edward',),
 ('Martha',),
 ('Aaron',),
 ('Ellie',),
 ('João',),
 ('Madalena',),
 ('Hannah',),
 ('Fynn',),
 ('Niklas',),
 ('Camille',),
 ('Dominique',),
 ('Marc',),
 ('Wyatt',),
 ('Isabelle',),
 ('Terhi',),
 ('Ladislav',),
 ('Hugh',),
 ('Lucas',),
 ('Johannes',),
 ('Stanisław',),
 ('Enrique',),
 ('Joakim',),
 ('Emma',),
 ('Phil',),
 ('Steve',),
 ('Mark',),
 ('Diego',),
 ('Luis',),
 ('Manoj',),
 ('Puja',)]

**3.3 Mostramos las columnas FirstName y LastName**

In [6]:
cursor.execute("SELECT FirstName, LastName FROM Customer")
firstName_LastName = cursor.fetchall()
firstName_LastName

[('Luís', 'Gonçalves'),
 ('Leonie', 'Köhler'),
 ('François', 'Tremblay'),
 ('Bjørn', 'Hansen'),
 ('František', 'Wichterlová'),
 ('Helena', 'Holý'),
 ('Astrid', 'Gruber'),
 ('Daan', 'Peeters'),
 ('Kara', 'Nielsen'),
 ('Eduardo', 'Martins'),
 ('Alexandre', 'Rocha'),
 ('Roberto', 'Almeida'),
 ('Fernanda', 'Ramos'),
 ('Mark', 'Philips'),
 ('Jennifer', 'Peterson'),
 ('Frank', 'Harris'),
 ('Jack', 'Smith'),
 ('Michelle', 'Brooks'),
 ('Tim', 'Goyer'),
 ('Dan', 'Miller'),
 ('Kathy', 'Chase'),
 ('Heather', 'Leacock'),
 ('John', 'Gordon'),
 ('Frank', 'Ralston'),
 ('Victor', 'Stevens'),
 ('Richard', 'Cunningham'),
 ('Patrick', 'Gray'),
 ('Julia', 'Barnett'),
 ('Robert', 'Brown'),
 ('Edward', 'Francis'),
 ('Martha', 'Silk'),
 ('Aaron', 'Mitchell'),
 ('Ellie', 'Sullivan'),
 ('João', 'Fernandes'),
 ('Madalena', 'Sampaio'),
 ('Hannah', 'Schneider'),
 ('Fynn', 'Zimmermann'),
 ('Niklas', 'Schröder'),
 ('Camille', 'Bernard'),
 ('Dominique', 'Lefebvre'),
 ('Marc', 'Dubois'),
 ('Wyatt', 'Girard'),
 ('Isab

**3.4 Mostramos las columnas Address, Phone, Fax**

In [7]:
cursor.execute("SELECT Address, Phone, Fax FROM Customer")
Addres_Phone_Fax = cursor.fetchall()
Addres_Phone_Fax

[('Av. Brigadeiro Faria Lima, 2170',
  '+55 (12) 3923-5555',
  '+55 (12) 3923-5566'),
 ('Theodor-Heuss-Straße 34', '+49 0711 2842222', None),
 ('1498 rue Bélanger', '+1 (514) 721-4711', None),
 ('Ullevålsveien 14', '+47 22 44 22 22', None),
 ('Klanova 9/506', '+420 2 4172 5555', '+420 2 4172 5555'),
 ('Rilská 3174/6', '+420 2 4177 0449', None),
 ('Rotenturmstraße 4, 1010 Innere Stadt', '+43 01 5134505', None),
 ('Grétrystraat 63', '+32 02 219 03 03', None),
 ('Sønder Boulevard 51', '+453 3331 9991', None),
 ('Rua Dr. Falcão Filho, 155', '+55 (11) 3033-5446', '+55 (11) 3033-4564'),
 ('Av. Paulista, 2022', '+55 (11) 3055-3278', '+55 (11) 3055-8131'),
 ('Praça Pio X, 119', '+55 (21) 2271-7000', '+55 (21) 2271-7070'),
 ('Qe 7 Bloco G', '+55 (61) 3363-5547', '+55 (61) 3363-7855'),
 ('8210 111 ST NW', '+1 (780) 434-4554', '+1 (780) 434-5565'),
 ('700 W Pender Street', '+1 (604) 688-2255', '+1 (604) 688-8756'),
 ('1600 Amphitheatre Parkway', '+1 (650) 253-0000', '+1 (650) 253-0000'),
 ('1 Mic

**3.5 Mostramos todas las columnas**

In [8]:
cursor.execute("SELECT * FROM Customer")
todas_columnas = cursor.fetchall()
todas_columnas

[(1,
  'Luís',
  'Gonçalves',
  'Embraer - Empresa Brasileira de Aeronáutica S.A.',
  'Av. Brigadeiro Faria Lima, 2170',
  'São José dos Campos',
  'SP',
  'Brazil',
  '12227-000',
  '+55 (12) 3923-5555',
  '+55 (12) 3923-5566',
  'luisg@embraer.com.br',
  3),
 (2,
  'Leonie',
  'Köhler',
  None,
  'Theodor-Heuss-Straße 34',
  'Stuttgart',
  None,
  'Germany',
  '70174',
  '+49 0711 2842222',
  None,
  'leonekohler@surfeu.de',
  5),
 (3,
  'François',
  'Tremblay',
  None,
  '1498 rue Bélanger',
  'Montréal',
  'QC',
  'Canada',
  'H2G 1A7',
  '+1 (514) 721-4711',
  None,
  'ftremblay@gmail.com',
  3),
 (4,
  'Bjørn',
  'Hansen',
  None,
  'Ullevålsveien 14',
  'Oslo',
  None,
  'Norway',
  '0171',
  '+47 22 44 22 22',
  None,
  'bjorn.hansen@yahoo.no',
  4),
 (5,
  'František',
  'Wichterlová',
  'JetBrains s.r.o.',
  'Klanova 9/506',
  'Prague',
  None,
  'Czech Republic',
  '14700',
  '+420 2 4172 5555',
  '+420 2 4172 5555',
  'frantisekw@jetbrains.com',
  4),
 (6,
  'Helena',
  'H

### 4 Ordenamos la base de datos

**4.1 Ordenamos por genero y mostramos los 6 primeros resultados**

In [9]:
cursor.execute("SELECT * FROM Genre ORDER BY Name ASC LIMIT 6")
genero = cursor.fetchall()
print("Los 6 primeros géneros ordenados alfabéticamente:")
for row in genero:
    print(row)

Los 6 primeros géneros ordenados alfabéticamente:
(23, 'Alternative')
(4, 'Alternative & Punk')
(6, 'Blues')
(11, 'Bossa Nova')
(24, 'Classical')
(22, 'Comedy')


**4.2 Mostramos las 2 ultimas filas de la tabla**

In [10]:
cursor.execute("SELECT * FROM Genre ORDER BY GenreId DESC LIMIT 2")
ultimasFilas = cursor.fetchall()
print("Últimas 2 filas de la tabla Genre:")
for row in ultimasFilas:
    print(row)

Últimas 2 filas de la tabla Genre:
(25, 'Opera')
(24, 'Classical')


### 5 En busca del peor cliente

**5.1 Usuarios que han reportado más de 3 incidencias**

In [11]:
cursor.execute("SELECT * FROM Customer WHERE SupportRepId > 3")
usuarios = cursor.fetchall()
print("Usuarios que han reportado más de 3 incidencias:\n")
for row in usuarios:
    print(row)

Usuarios que han reportado más de 3 incidencias:

(4, 'Bjørn', 'Hansen', None, 'Ullevålsveien 14', 'Oslo', None, 'Norway', '0171', '+47 22 44 22 22', None, 'bjorn.hansen@yahoo.no', 4)
(5, 'František', 'Wichterlová', 'JetBrains s.r.o.', 'Klanova 9/506', 'Prague', None, 'Czech Republic', '14700', '+420 2 4172 5555', '+420 2 4172 5555', 'frantisekw@jetbrains.com', 4)
(8, 'Daan', 'Peeters', None, 'Grétrystraat 63', 'Brussels', None, 'Belgium', '1000', '+32 02 219 03 03', None, 'daan_peeters@apple.be', 4)
(9, 'Kara', 'Nielsen', None, 'Sønder Boulevard 51', 'Copenhagen', None, 'Denmark', '1720', '+453 3331 9991', None, 'kara.nielsen@jubii.dk', 4)
(10, 'Eduardo', 'Martins', 'Woodstock Discos', 'Rua Dr. Falcão Filho, 155', 'São Paulo', 'SP', 'Brazil', '01007-010', '+55 (11) 3033-5446', '+55 (11) 3033-4564', 'eduardo@woodstock.com.br', 4)
(13, 'Fernanda', 'Ramos', None, 'Qe 7 Bloco G', 'Brasília', 'DF', 'Brazil', '71020-677', '+55 (61) 3363-5547', '+55 (61) 3363-7855', 'fernadaramos4@uol.com.br

**5.2 Usuarios que son de Brasil**

In [12]:
cursor.execute("SELECT * FROM Customer WHERE Country = 'Brazil'")
Brasil = cursor.fetchall()
print("Usuarios que viven en Brasil:\n")
for row in Brasil:
    print(row)

Usuarios que viven en Brasil:

(1, 'Luís', 'Gonçalves', 'Embraer - Empresa Brasileira de Aeronáutica S.A.', 'Av. Brigadeiro Faria Lima, 2170', 'São José dos Campos', 'SP', 'Brazil', '12227-000', '+55 (12) 3923-5555', '+55 (12) 3923-5566', 'luisg@embraer.com.br', 3)
(10, 'Eduardo', 'Martins', 'Woodstock Discos', 'Rua Dr. Falcão Filho, 155', 'São Paulo', 'SP', 'Brazil', '01007-010', '+55 (11) 3033-5446', '+55 (11) 3033-4564', 'eduardo@woodstock.com.br', 4)
(11, 'Alexandre', 'Rocha', 'Banco do Brasil S.A.', 'Av. Paulista, 2022', 'São Paulo', 'SP', 'Brazil', '01310-200', '+55 (11) 3055-3278', '+55 (11) 3055-8131', 'alero@uol.com.br', 5)
(12, 'Roberto', 'Almeida', 'Riotur', 'Praça Pio X, 119', 'Rio de Janeiro', 'RJ', 'Brazil', '20040-020', '+55 (21) 2271-7000', '+55 (21) 2271-7070', 'roberto.almeida@riotur.gov.br', 3)
(13, 'Fernanda', 'Ramos', None, 'Qe 7 Bloco G', 'Brasília', 'DF', 'Brazil', '71020-677', '+55 (61) 3363-5547', '+55 (61) 3363-7855', 'fernadaramos4@uol.com.br', 4)


**5.3 Usuarios cuyo codigo postal empieza por 7**

In [13]:
cursor.execute("SELECT * FROM Customer WHERE PostalCode LIKE '7%'")
codigoPostal = cursor.fetchall()
print("Usuarios con código postal empezado en 7:\n")
for row in codigoPostal:
    print(row)

Usuarios con código postal empezado en 7:

(2, 'Leonie', 'Köhler', None, 'Theodor-Heuss-Straße 34', 'Stuttgart', None, 'Germany', '70174', '+49 0711 2842222', None, 'leonekohler@surfeu.de', 5)
(13, 'Fernanda', 'Ramos', None, 'Qe 7 Bloco G', 'Brasília', 'DF', 'Brazil', '71020-677', '+55 (61) 3363-5547', '+55 (61) 3363-7855', 'fernadaramos4@uol.com.br', 4)
(26, 'Richard', 'Cunningham', None, '2211 W Berry Street', 'Fort Worth', 'TX', 'USA', '76110', '+1 (817) 924-7272', None, 'ricunningham@hotmail.com', 4)
(39, 'Camille', 'Bernard', None, '4, Rue Milton', 'Paris', None, 'France', '75009', '+33 01 49 70 65 65', None, 'camille.bernard@yahoo.fr', 4)
(40, 'Dominique', 'Lefebvre', None, '8, Rue Hanovre', 'Paris', None, 'France', '75002', '+33 01 47 42 71 71', None, 'dominiquelefebvre@gmail.com', 4)


**5.4 Usuarios que usan email de hotmail**

In [14]:
cursor.execute("SELECT * FROM Customer WHERE Email LIKE '%@hotmail.%'")
hotmail = cursor.fetchall()
print("Usuarios con email de hotmail:\n")
for row in hotmail:
    print(row)

Usuarios con email de hotmail:

(21, 'Kathy', 'Chase', None, '801 W 4th Street', 'Reno', 'NV', 'USA', '89503', '+1 (775) 223-7665', None, 'kachase@hotmail.com', 5)
(26, 'Richard', 'Cunningham', None, '2211 W Berry Street', 'Fort Worth', 'TX', 'USA', '76110', '+1 (817) 924-7272', None, 'ricunningham@hotmail.com', 4)
(41, 'Marc', 'Dubois', None, '11, Place Bellecour', 'Lyon', None, 'France', '69002', '+33 04 78 30 30 30', None, 'marc.dubois@hotmail.com', 5)
(52, 'Emma', 'Jones', None, '202 Hoxton Street', 'London', None, 'United Kingdom', 'N1 5LH', '+44 020 7707 0707', None, 'emma_jones@hotmail.com', 3)


**5.5 Usuarios nacidos en EEUU o Canada**

In [15]:
cursor.execute("SELECT * FROM Customer WHERE Country IN ('USA', 'Canada')")
USA_Canada = cursor.fetchall()
print("Usuarios nacidos en Estados Unidos o Canadá:\n")
for row in USA_Canada:
    print(row)

Usuarios nacidos en Estados Unidos o Canadá:

(3, 'François', 'Tremblay', None, '1498 rue Bélanger', 'Montréal', 'QC', 'Canada', 'H2G 1A7', '+1 (514) 721-4711', None, 'ftremblay@gmail.com', 3)
(14, 'Mark', 'Philips', 'Telus', '8210 111 ST NW', 'Edmonton', 'AB', 'Canada', 'T6G 2C7', '+1 (780) 434-4554', '+1 (780) 434-5565', 'mphilips12@shaw.ca', 5)
(15, 'Jennifer', 'Peterson', 'Rogers Canada', '700 W Pender Street', 'Vancouver', 'BC', 'Canada', 'V6C 1G8', '+1 (604) 688-2255', '+1 (604) 688-8756', 'jenniferp@rogers.ca', 3)
(16, 'Frank', 'Harris', 'Google Inc.', '1600 Amphitheatre Parkway', 'Mountain View', 'CA', 'USA', '94043-1351', '+1 (650) 253-0000', '+1 (650) 253-0000', 'fharris@google.com', 4)
(17, 'Jack', 'Smith', 'Microsoft Corporation', '1 Microsoft Way', 'Redmond', 'WA', 'USA', '98052-8300', '+1 (425) 882-8080', '+1 (425) 882-8081', 'jacksmith@microsoft.com', 5)
(18, 'Michelle', 'Brooks', None, '627 Broadway', 'New York', 'NY', 'USA', '10012-2612', '+1 (212) 221-3546', '+1 (212)

**5.6 Usuarios que usan Gmail**

In [16]:
cursor.execute("SELECT * FROM Customer WHERE Email LIKE '%@gmail.%'")
gmail = cursor.fetchall()
print("Usuarios con email de gmail:\n")
for row in gmail:
    print(row)

Usuarios con email de gmail:

(3, 'François', 'Tremblay', None, '1498 rue Bélanger', 'Montréal', 'QC', 'Canada', 'H2G 1A7', '+1 (514) 721-4711', None, 'ftremblay@gmail.com', 3)
(6, 'Helena', 'Holý', None, 'Rilská 3174/6', 'Prague', None, 'Czech Republic', '14300', '+420 2 4177 0449', None, 'hholy@gmail.com', 5)
(22, 'Heather', 'Leacock', None, '120 S Orange Ave', 'Orlando', 'FL', 'USA', '32801', '+1 (407) 999-7788', None, 'hleacock@gmail.com', 4)
(24, 'Frank', 'Ralston', None, '162 E Superior Street', 'Chicago', 'IL', 'USA', '60611', '+1 (312) 332-3232', None, 'fralston@gmail.com', 3)
(28, 'Julia', 'Barnett', None, '302 S 700 E', 'Salt Lake City', 'UT', 'USA', '84102', '+1 (801) 531-7272', None, 'jubarnett@gmail.com', 5)
(31, 'Martha', 'Silk', None, '194A Chain Lake Drive', 'Halifax', 'NS', 'Canada', 'B3S 1C5', '+1 (902) 450-0450', None, 'marthasilk@gmail.com', 5)
(40, 'Dominique', 'Lefebvre', None, '8, Rue Hanovre', 'Paris', None, 'France', '75002', '+33 01 47 42 71 71', None, 'domini

**5.7 Usuarios que trabajan en Apple**

In [17]:
cursor.execute("SELECT * FROM Customer WHERE Company LIKE '%Apple%'")
apple = cursor.fetchall()
print("Usuario que trabaja en Apple:\n")
for row in apple:
    print(row)

Usuario que trabaja en Apple:

(19, 'Tim', 'Goyer', 'Apple Inc.', '1 Infinite Loop', 'Cupertino', 'CA', 'USA', '95014', '+1 (408) 996-1010', '+1 (408) 996-1011', 'tgoyer@apple.com', 3)


**5.8 Usuarios que han reportado entre 3 y 4 incidencias**

In [18]:
cursor.execute("SELECT * FROM Customer WHERE SupportRepId BETWEEN 3 AND 4")
incidencias = cursor.fetchall()
print("Usuarios que han reportado entre 3 y 4 incidencias:\n")
for row in incidencias:
    print(row)

Usuarios que han reportado entre 3 y 4 incidencias:

(1, 'Luís', 'Gonçalves', 'Embraer - Empresa Brasileira de Aeronáutica S.A.', 'Av. Brigadeiro Faria Lima, 2170', 'São José dos Campos', 'SP', 'Brazil', '12227-000', '+55 (12) 3923-5555', '+55 (12) 3923-5566', 'luisg@embraer.com.br', 3)
(3, 'François', 'Tremblay', None, '1498 rue Bélanger', 'Montréal', 'QC', 'Canada', 'H2G 1A7', '+1 (514) 721-4711', None, 'ftremblay@gmail.com', 3)
(12, 'Roberto', 'Almeida', 'Riotur', 'Praça Pio X, 119', 'Rio de Janeiro', 'RJ', 'Brazil', '20040-020', '+55 (21) 2271-7000', '+55 (21) 2271-7070', 'roberto.almeida@riotur.gov.br', 3)
(15, 'Jennifer', 'Peterson', 'Rogers Canada', '700 W Pender Street', 'Vancouver', 'BC', 'Canada', 'V6C 1G8', '+1 (604) 688-2255', '+1 (604) 688-8756', 'jenniferp@rogers.ca', 3)
(18, 'Michelle', 'Brooks', None, '627 Broadway', 'New York', 'NY', 'USA', '10012-2612', '+1 (212) 221-3546', '+1 (212) 221-4679', 'michelleb@aol.com', 3)
(19, 'Tim', 'Goyer', 'Apple Inc.', '1 Infinite Loo

### 6 Empleados

**6.1 Mostramos la cantidad de empleados por ciudad**

In [19]:
cursor.execute("SELECT City, COUNT(*) FROM Employee GROUP BY City")
cantidad = cursor.fetchall()
print("Cantidad de empleados por ciudad:\n")
for row in cantidad:
    print(row)

Cantidad de empleados por ciudad:

('Calgary', 5)
('Edmonton', 1)
('Lethbridge', 2)


**6.2 Empleados por departamento**

In [20]:
cursor.execute("SELECT City, COUNT(*) FROM Employee GROUP BY PostalCode")
departamento = cursor.fetchall()
print("Cantidad de empleados por departamento:\n")
for row in departamento:
    print(row)

Cantidad de empleados por departamento:

('Lethbridge', 1)
('Lethbridge', 1)
('Calgary', 1)
('Calgary', 1)
('Calgary', 1)
('Calgary', 1)
('Calgary', 1)
('Edmonton', 1)


**6.3 Edad de los empleados**

In [21]:
cursor.execute("SELECT FirstName, LastName, BirthDate FROM Employee")
edad = cursor.fetchall()
print("Edad de cada empleado:\n")
for row in edad:
    first_name, last_name, birth_date = row
    print(f"{first_name} {last_name}: {birth_date}")


Edad de cada empleado:

Andrew Adams: 1962-02-18 00:00:00
Nancy Edwards: 1958-12-08 00:00:00
Jane Peacock: 1973-08-29 00:00:00
Margaret Park: 1947-09-19 00:00:00
Steve Johnson: 1965-03-03 00:00:00
Michael Mitchell: 1973-07-01 00:00:00
Robert King: 1970-05-29 00:00:00
Laura Callahan: 1968-01-09 00:00:00


**6.4 Epleados contratados de media por año**

In [22]:
cursor.execute("SELECT AVG(strftime('%Y', HireDate)) - strftime('%Y', BirthDate) FROM Employee")
contratados = cursor.fetchone()
print("Empleados contratados, de media, por año:\n")
print(contratados[0])

Empleados contratados, de media, por año:

40.875


### 7 Experto en canciones

**7.1 Mostramos todas las canciones con el MediaType Protected AAC audio file**

In [23]:
cursor.execute("SELECT * FROM Track WHERE MediaTypeId IN (SELECT MediaTypeId FROM MediaType WHERE Name = 'Protected AAC audio file')")
mediaType = cursor.fetchall()
print("Canciones con MediaType Protected AAC audio file:\n")
for row in mediaType:
    print(row)

Canciones con MediaType Protected AAC audio file:

(2, 'Balls to the Wall', 2, 2, 1, None, 342562, 5510424, 0.99)
(3, 'Fast As a Shark', 3, 2, 1, 'F. Baltes, S. Kaufman, U. Dirkscneider & W. Hoffman', 230619, 3990994, 0.99)
(4, 'Restless and Wild', 3, 2, 1, 'F. Baltes, R.A. Smith-Diesel, S. Kaufman, U. Dirkscneider & W. Hoffman', 252051, 4331779, 0.99)
(5, 'Princess of the Dawn', 3, 2, 1, 'Deaffy & R.A. Smith-Diesel', 375418, 6290521, 0.99)
(1146, 'Welcome to the Jungle', 90, 2, 1, None, 273552, 4538451, 0.99)
(1147, "It's So Easy", 90, 2, 1, None, 202824, 3394019, 0.99)
(1148, 'Nightrain', 90, 2, 1, None, 268537, 4457283, 0.99)
(1149, 'Out Ta Get Me', 90, 2, 1, None, 263893, 4382147, 0.99)
(1150, 'Mr. Brownstone', 90, 2, 1, None, 228924, 3816323, 0.99)
(1151, 'Paradise City', 90, 2, 1, None, 406347, 6687123, 0.99)
(1152, 'My Michelle', 90, 2, 1, None, 219961, 3671299, 0.99)
(1153, 'Think About You', 90, 2, 1, None, 231640, 3860275, 0.99)
(1154, "Sweet Child O' Mine", 90, 2, 1, None, 3

**7.2 Mostramos canciones con algún MediaType con AAC**

In [24]:
cursor.execute("SELECT * FROM Track WHERE MediaTypeId IN (SELECT MediaTypeId FROM MediaType WHERE Name LIKE '%AAC%')")
mediaTypeAac = cursor.fetchall()
print("Canciones con algún MediaType con AAC:\n")
for row in mediaTypeAac:
    print(row)

Canciones con algún MediaType con AAC:

(2, 'Balls to the Wall', 2, 2, 1, None, 342562, 5510424, 0.99)
(3, 'Fast As a Shark', 3, 2, 1, 'F. Baltes, S. Kaufman, U. Dirkscneider & W. Hoffman', 230619, 3990994, 0.99)
(4, 'Restless and Wild', 3, 2, 1, 'F. Baltes, R.A. Smith-Diesel, S. Kaufman, U. Dirkscneider & W. Hoffman', 252051, 4331779, 0.99)
(5, 'Princess of the Dawn', 3, 2, 1, 'Deaffy & R.A. Smith-Diesel', 375418, 6290521, 0.99)
(1146, 'Welcome to the Jungle', 90, 2, 1, None, 273552, 4538451, 0.99)
(1147, "It's So Easy", 90, 2, 1, None, 202824, 3394019, 0.99)
(1148, 'Nightrain', 90, 2, 1, None, 268537, 4457283, 0.99)
(1149, 'Out Ta Get Me', 90, 2, 1, None, 263893, 4382147, 0.99)
(1150, 'Mr. Brownstone', 90, 2, 1, None, 228924, 3816323, 0.99)
(1151, 'Paradise City', 90, 2, 1, None, 406347, 6687123, 0.99)
(1152, 'My Michelle', 90, 2, 1, None, 219961, 3671299, 0.99)
(1153, 'Think About You', 90, 2, 1, None, 231640, 3860275, 0.99)
(1154, "Sweet Child O' Mine", 90, 2, 1, None, 356424, 5879

**7.3 Canciones que duran más de 2 minutos**

In [25]:
cursor.execute("SELECT * FROM Track WHERE Milliseconds > 120000")
dosMinutos = cursor.fetchall()
print("Canciones que duran más de 2 minutos:\n")
for row in dosMinutos:
    print(row)

Canciones que duran más de 2 minutos:

(1, 'For Those About To Rock (We Salute You)', 1, 1, 1, 'Angus Young, Malcolm Young, Brian Johnson', 343719, 11170334, 0.99)
(2, 'Balls to the Wall', 2, 2, 1, None, 342562, 5510424, 0.99)
(3, 'Fast As a Shark', 3, 2, 1, 'F. Baltes, S. Kaufman, U. Dirkscneider & W. Hoffman', 230619, 3990994, 0.99)
(4, 'Restless and Wild', 3, 2, 1, 'F. Baltes, R.A. Smith-Diesel, S. Kaufman, U. Dirkscneider & W. Hoffman', 252051, 4331779, 0.99)
(5, 'Princess of the Dawn', 3, 2, 1, 'Deaffy & R.A. Smith-Diesel', 375418, 6290521, 0.99)
(6, 'Put The Finger On You', 1, 1, 1, 'Angus Young, Malcolm Young, Brian Johnson', 205662, 6713451, 0.99)
(7, "Let's Get It Up", 1, 1, 1, 'Angus Young, Malcolm Young, Brian Johnson', 233926, 7636561, 0.99)
(8, 'Inject The Venom', 1, 1, 1, 'Angus Young, Malcolm Young, Brian Johnson', 210834, 6852860, 0.99)
(9, 'Snowballed', 1, 1, 1, 'Angus Young, Malcolm Young, Brian Johnson', 203102, 6599424, 0.99)
(10, 'Evil Walks', 1, 1, 1, 'Angus Young

**7.4 Caciones de Jazz**

In [26]:
cursor.execute("SELECT * FROM Track WHERE GenreId IN (SELECT GenreId FROM Genre WHERE Name = 'Jazz')")
jazz = cursor.fetchall()
print("Canciones de Jazz:\n")
for row in jazz:
    print(row)

Canciones de Jazz:

(63, 'Desafinado', 8, 1, 2, None, 185338, 5990473, 0.99)
(64, 'Garota De Ipanema', 8, 1, 2, None, 285048, 9348428, 0.99)
(65, 'Samba De Uma Nota Só (One Note Samba)', 8, 1, 2, None, 137273, 4535401, 0.99)
(66, 'Por Causa De Você', 8, 1, 2, None, 169900, 5536496, 0.99)
(67, 'Ligia', 8, 1, 2, None, 251977, 8226934, 0.99)
(68, 'Fotografia', 8, 1, 2, None, 129227, 4198774, 0.99)
(69, 'Dindi (Dindi)', 8, 1, 2, None, 253178, 8149148, 0.99)
(70, 'Se Todos Fossem Iguais A Você (Instrumental)', 8, 1, 2, None, 134948, 4393377, 0.99)
(71, 'Falando De Amor', 8, 1, 2, None, 219663, 7121735, 0.99)
(72, 'Angela', 8, 1, 2, None, 169508, 5574957, 0.99)
(73, 'Corcovado (Quiet Nights Of Quiet Stars)', 8, 1, 2, None, 205662, 6687994, 0.99)
(74, 'Outra Vez', 8, 1, 2, None, 126511, 4110053, 0.99)
(75, 'O Boto (Bôto)', 8, 1, 2, None, 366837, 12089673, 0.99)
(76, 'Canta, Canta Mais', 8, 1, 2, None, 271856, 8719426, 0.99)
(123, 'Quadrant', 13, 1, 2, 'Billy Cobham', 261851, 8538199, 0.99)
(1

**7.5 Canción más pesada**

In [27]:
cursor.execute("SELECT * FROM Track WHERE Bytes = (SELECT MAX(Bytes) FROM Track)")
pesada = cursor.fetchone()
print("Canción más pesada:\n")
pesada

Canción más pesada:



(3224, 'Through a Looking Glass', 229, 3, 21, None, 5088838, 1059546140, 1.99)

**7.6 ¿Cuantos discos hay de Led Zeppelin?**

In [28]:
cursor.execute("SELECT COUNT(*) FROM Album WHERE ArtistId IN (SELECT ArtistId FROM Artist WHERE Name = 'Led Zeppelin')")
ledZeppelin = cursor.fetchone()
print(f"Led Zeppelin tiene {ledZeppelin[0]} discos")

Led Zeppelin tiene 14 discos


**7.7 Cúal es el precio del disco Houses Of The Holy de Led Zeppelin**

In [29]:
cursor.execute("SELECT UnitPrice FROM Track WHERE AlbumId IN (SELECT AlbumId FROM Album WHERE Title = 'Houses Of The Holy' AND ArtistId IN (SELECT ArtistId FROM Artist WHERE Name = 'Led Zeppelin'))")
precio = cursor.fetchone()
print(f"El precio del disco Houses Of The Holy de Led Zeppelin son {precio[0]}")

El precio del disco Houses Of The Holy de Led Zeppelin son 0.99


### 8 Compañias

**8.1 Creamos una vista llamada  Customer_with_companies, donde estarán incluidos todos los resultados salvo cuando Company sea NULL**

In [30]:
cursor.execute("CREATE VIEW IF NOT EXISTS Customer_with_companies AS \
                SELECT * FROM Customer WHERE Company IS NOT NULL")

**8.2 Ordenamos los resultados por orden alfabetico**

In [31]:
cursor.execute("SELECT * FROM Customer_with_companies ORDER BY Company")
orden_alfabetico = cursor.fetchall()
print("Resultados ordenados por orden alfabético de Company:\n")
for row in orden_alfabetico:
    print(row)

Resultados ordenados por orden alfabético de Company:

(19, 'Tim', 'Goyer', 'Apple Inc.', '1 Infinite Loop', 'Cupertino', 'CA', 'USA', '95014', '+1 (408) 996-1010', '+1 (408) 996-1011', 'tgoyer@apple.com', 3)
(11, 'Alexandre', 'Rocha', 'Banco do Brasil S.A.', 'Av. Paulista, 2022', 'São Paulo', 'SP', 'Brazil', '01310-200', '+55 (11) 3055-3278', '+55 (11) 3055-8131', 'alero@uol.com.br', 5)
(1, 'Luís', 'Gonçalves', 'Embraer - Empresa Brasileira de Aeronáutica S.A.', 'Av. Brigadeiro Faria Lima, 2170', 'São José dos Campos', 'SP', 'Brazil', '12227-000', '+55 (12) 3923-5555', '+55 (12) 3923-5566', 'luisg@embraer.com.br', 3)
(16, 'Frank', 'Harris', 'Google Inc.', '1600 Amphitheatre Parkway', 'Mountain View', 'CA', 'USA', '94043-1351', '+1 (650) 253-0000', '+1 (650) 253-0000', 'fharris@google.com', 4)
(5, 'František', 'Wichterlová', 'JetBrains s.r.o.', 'Klanova 9/506', 'Prague', None, 'Czech Republic', '14700', '+420 2 4172 5555', '+420 2 4172 5555', 'frantisekw@jetbrains.com', 4)
(17, 'Jack',

**8.3 Compañias de Brasil**

In [32]:
cursor.execute("SELECT DISTINCT Company FROM Customer_with_companies WHERE Country = 'Brazil'")
Brasil = cursor.fetchall()
print("Compañías de Brasil:\n")
for row in Brasil:
    print(row[0])

Compañías de Brasil:

Embraer - Empresa Brasileira de Aeronáutica S.A.
Woodstock Discos
Banco do Brasil S.A.
Riotur


### 9 Facturas

**9.1 Titulo de la cancion que menos pesa**

In [33]:
cursor.execute("SELECT Name FROM Track WHERE Bytes = (SELECT MIN(Bytes) FROM Track)")
cancion = cursor.fetchone()
print("Título de la canción que menos pesa:\n")
print(cancion[0])

Título de la canción que menos pesa:

É Uma Partida De Futebol


**9.2 Titulo de la cancion que mas dura**

In [34]:
cursor.execute("SELECT Name FROM Track WHERE Milliseconds = (SELECT MAX(Milliseconds) FROM Track)")
duracion = cursor.fetchone()
print("Título de la canción que más dura:\n")
print(duracion[0])

Título de la canción que más dura:

Occupation / Precipice


**9.3 Canciones que cuestan 1$ o más**

In [35]:
cursor.execute("SELECT COUNT(*) FROM Track WHERE UnitPrice >= 1")
unDolar = cursor.fetchone()
print("Cantidad de canciones que cuestan 1$ o más:\ng")
print(unDolar[0])

Cantidad de canciones que cuestan 1$ o más:
g
213


**9.4 Cantidad de canciones de Queen**

In [6]:
cursor.execute("SELECT COUNT (DISTINCT TrackId) FROM Track WHERE Composer LIKE '%Queen%'")
Queen = cursor.fetchone()
print("Cantidad de canciones de Queen:\n")
print(Queen[0])

Cantidad de canciones de Queen:

10


**9.5 Media de duración de todas las canciones**

In [37]:
cursor.execute("SELECT AVG(Milliseconds) FROM Track")
duracion = cursor.fetchone()
print("Media de duración entre todas las canciones:\n")
print(round(duracion[0]))

Media de duración entre todas las canciones:

393599


**9.6 Media de peso entre todas las canciones de U2**

In [38]:
cursor.execute("SELECT AVG(Bytes) FROM Track WHERE AlbumId IN (SELECT AlbumId FROM Album WHERE ArtistId IN (SELECT ArtistId FROM Artist WHERE Name = 'U2'))")
U2 = cursor.fetchone()
print("Media de peso entre todas las canciones de U2:\n")
print(round(U2[0]))

Media de peso entre todas las canciones de U2:

7978790


**9.7 Cantidad de canciones en las que Bill Berry es compositor**

In [39]:
cursor.execute("SELECT COUNT(*) FROM Track WHERE Composer = 'Bill Berry'")
Bill = cursor.fetchone()
print("Cantidad de canciones en las que Bill Berry es el compositor:\n")
print(Bill[0])

Cantidad de canciones en las que Bill Berry es el compositor:

0


**9.8 Conversión de Bytes a MB de todas las canciones**

In [40]:
cursor.execute("SELECT Name, Bytes / 1024.0 / 1024.0 AS Mb FROM Track")
MB = cursor.fetchall()
print("Tracks con columna Bytes convertida a Mb:\n")
for row in MB:
    print(row)

Tracks con columna Bytes convertida a Mb:

('For Those About To Rock (We Salute You)', 10.652860641479492)
('Balls to the Wall', 5.255149841308594)
('Fast As a Shark', 3.8061084747314453)
('Restless and Wild', 4.131106376647949)
('Princess of the Dawn', 5.99910831451416)
('Put The Finger On You', 6.4024457931518555)
("Let's Get It Up", 7.282792091369629)
('Inject The Venom', 6.535396575927734)
('Snowballed', 6.293701171875)
('Evil Walks', 8.212323188781738)
('C.O.D.', 6.262125015258789)
('Breaking The Rules', 8.198585510253906)
('Night Of The Long Knives', 6.3956708908081055)
('Spellbound', 8.40858268737793)
('Go Down', 10.345088005065918)
('Dog Eat Dog', 6.706392288208008)
('Let There Be Rock', 11.464367866516113)
('Bad Boy Boogie', 8.369579315185547)
('Problem Child', 10.12527084350586)
('Overdose', 11.507314682006836)
("Hell Ain't A Bad Place To Be", 7.945333480834961)
('Whole Lotta Rosie', 10.058549880981445)
('Walk On Water', 9.269312858581543)
('Love In An Elevator', 10.063220024

('Você', 7.545186996459961)
('Não Quero Dinheiro', 4.7986955642700195)
('Eu Amo Você', 7.547977447509766)
('A Festa Do Santo Reis', 4.963870048522949)
("I Don't Know What To Do With Myself", 6.915548324584961)
('Padre Cícero', 4.3694353103637695)
('Nosso Adeus', 6.478567123413086)
('Canário Do Reino', 4.341943740844727)
('Preciso Ser Amado', 5.358595848083496)
('Balanço', 6.571127891540527)
('Preciso Aprender A Ser Só', 4.972356796264648)
('Esta É A Canção', 5.788739204406738)
('Formigueiro', 8.06344223022461)
('Comida', 10.286882400512695)
('Go Back', 7.313632011413574)
('Prá Dizer Adeus', 7.040069580078125)
('Família', 6.930788040161133)
('Os Cegos Do Castelo', 9.411036491394043)
('O Pulso', 6.262777328491211)
('Marvin', 8.336490631103516)
('Nem 5 Minutos Guardados', 7.7665300369262695)
('Flores', 6.816880226135254)
('Palavras', 5.040850639343262)
('Hereditário', 4.787966728210449)
('A Melhor Forma', 6.05577278137207)
('Cabeça Dinossauro', 1.1643695831298828)
('32 Dentes', 5.87263488

### 10 Buscando Facturas

**10.1 Muestra: InvoiceId, nombre del cliente y BillingCountry**

In [41]:
cursor.execute("SELECT InvoiceId, Customer.FirstName || ' ' || Customer.LastName AS CustomerName, BillingCountry \
                FROM Invoice \
                INNER JOIN Customer ON Invoice.CustomerId = Customer.CustomerId \
                ORDER BY Total DESC")
resultado = cursor.fetchall()
print("InvoiceId, nombre del cliente y BillingCountry ordenados de mayor a menor por Total:\n")
for row in resultado:
    print(row)

InvoiceId, nombre del cliente y BillingCountry ordenados de mayor a menor por Total:

(404, 'Helena Holý', 'Czech Republic')
(299, 'Richard Cunningham', 'USA')
(96, 'Ladislav Kovács', 'Hungary')
(194, "Hugh O'Reilly", 'Ireland')
(89, 'Astrid Gruber', 'Austria')
(201, 'Victor Stevens', 'USA')
(88, 'Luis Rojas', 'Chile')
(306, 'František Wichterlová', 'Czech Republic')
(313, 'Isabelle Mercier', 'France')
(103, 'Frank Ralston', 'USA')
(208, 'Bjørn Hansen', 'Norway')
(193, 'Fynn Zimmermann', 'Germany')
(5, 'John Gordon', 'USA')
(12, 'Leonie Köhler', 'Germany')
(19, 'Dominique Lefebvre', 'France')
(26, 'Tim Goyer', 'USA')
(33, 'Luis Rojas', 'Chile')
(40, 'Hannah Schneider', 'Germany')
(47, 'Jennifer Peterson', 'Canada')
(54, 'Phil Hughes', 'United Kingdom')
(61, 'Aaron Mitchell', 'Canada')
(68, 'Alexandre Rocha', 'Brazil')
(75, 'Stanisław Wójcik', 'Poland')
(82, 'Julia Barnett', 'USA')
(110, 'François Tremblay', 'Canada')
(117, 'Marc Dubois', 'France')
(124, 'Dan Miller', 'USA')
(131, 'Mano

**10.2 Pais que más a facturado**

In [42]:
cursor.execute("SELECT BillingCountry, SUM(Total) AS TotalSales \
                FROM Invoice \
                GROUP BY BillingCountry \
                ORDER BY TotalSales DESC \
                LIMIT 1")
facturado = cursor.fetchone()
print("País que más ha facturado:\n")
print(facturado[0])

País que más ha facturado:

USA


### 11 Alimentando la productora

**11.1 Añade 5 artistas a la tabla Artist**

In [43]:
artistas = ['Grupo1', 'Grupo2', 'Grupo3', 'Grupo4', 'Grupo5']
for artista in artistas:
    cursor.execute("INSERT INTO Artist (Name) VALUES (?)", (artista,))
conexion.commit()
print("Se han añadido 5 artistas a la tabla Artist.")

Se han añadido 5 artistas a la tabla Artist.


**11.2 Introducor el MediaType Wav**

In [44]:
cursor.execute("INSERT INTO MediaType (Name) VALUES ('Wav')")
conexion.commit()
print("\nSe ha introducido el MediaType Wav.")


Se ha introducido el MediaType Wav.


**11.3 Obtener los 5 ultimos grupos añadidos**

In [45]:
cursor.execute("SELECT ArtistId, Name FROM Artist ORDER BY ArtistId DESC LIMIT 5")
grupos = cursor.fetchall()
ultimos_artistas = [row[0] for row in grupos]
print("Últimos 5 artistas añadidos:\n")
for row in grupos:
    print(row)

Últimos 5 artistas añadidos:

(280, 'Grupo5')
(279, 'Grupo4')
(278, 'Grupo3')
(277, 'Grupo2')
(276, 'Grupo1')


**11.4 Creamos 2 discos relacionados con los grupos**

In [46]:
album1 = ("Disco1", ultimos_artistas[0])
album2 = ("Disco2", ultimos_artistas[1])
cursor.execute("INSERT INTO Album (Title, ArtistId) VALUES (?, ?)", album1)
cursor.execute("INSERT INTO Album (Title, ArtistId) VALUES (?, ?)", album2)
conexion.commit()
print("Se han creado 2 discos relacionados con los grupos.\n")

Se han creado 2 discos relacionados con los grupos.



### 12 Artistas Desaparecidos

**12.1 Borrar U2**

In [47]:
cursor.execute("DELETE FROM Artist WHERE Name = 'U2'")
conexion.commit()
print("Se ha borrado a U2 de la tabla Artist.")

Se ha borrado a U2 de la tabla Artist.


**12.2 Borrar artistas que en su nombre contengan el simbolo &**

In [48]:
cursor.execute("DELETE FROM Artist WHERE Name LIKE '%&%'")
conexion.commit()
print("Se han borrado los artistas con el símbolo & en su nombre.")

Se han borrado los artistas con el símbolo & en su nombre.


**12.3 Borrar los artistas con las entre las Id 201 y 230**

In [49]:
cursor.execute("DELETE FROM Artist WHERE ArtistId BETWEEN 201 AND 230")
conexion.commit()
print("Se han borrado los artistas con una Id entre 201 y 230.")

Se han borrado los artistas con una Id entre 201 y 230.


**12.4 Borrar toda la tabla Track**

In [50]:
cursor.execute("DELETE FROM Track")
conexion.commit()
print("Se ha borrado toda la tabla Track.")

Se ha borrado toda la tabla Track.


### 13 Se acabaron las rebajas

**13.1 Subir el precio de las canciones de 0.99 a 2.99**

In [51]:
cursor.execute("UPDATE Track SET UnitPrice = 2.99 WHERE UnitPrice = 0.99")
conexion.commit()
print("Se ha actualizado el precio de todas las canciones de 0.99 a 2.99.")

Se ha actualizado el precio de todas las canciones de 0.99 a 2.99.


**13.2 Subir el precio de 1.99 a 4.99**

In [52]:
cursor.execute("UPDATE Track SET UnitPrice = 4.99 WHERE UnitPrice = 1.99")
conexion.commit()
print("Se ha actualizado el precio de todas las canciones de 1.99 a 4.99.")

Se ha actualizado el precio de todas las canciones de 1.99 a 4.99.


**13.3 Cambiar el Compositos 'Composer' por 'Sara' del TrackId 2000**

In [53]:
cursor.execute("UPDATE Track SET Composer = 'Sara' WHERE TrackId = 2000")
conexion.commit()
print("Se ha cambiado el Compositor (Composer) por 'Sara' del TrackId número 2000.")

Se ha cambiado el Compositor (Composer) por 'Sara' del TrackId número 2000.


**13.4 Cambiar el Compositor (Composer) por clasico de todas las canciones que empiecen con el nombre Concerto**

In [54]:
cursor.execute("UPDATE Track SET Composer = 'clasico' WHERE Name LIKE 'Concerto%'")
conexion.commit()
print("Se ha cambiado el Compositor (Composer) por 'clasico' de todas las canciones que empiecen con el nombre 'Concerto'.")

Se ha cambiado el Compositor (Composer) por 'clasico' de todas las canciones que empiecen con el nombre 'Concerto'.


### Cerramos la conexión con la base de datos

In [55]:
conexion.close()